# Aula 3

Como já vimos na aula anterior, a função `input()` é usada quando queremos receber os dados ou valores fornecidos pelo usuário através do teclado. 

No entanto, o valor que o usuário fornece e que será retornada pelo `input()` é sempre uma **string** e não um número. A função `primeira()` definida a seguir, exemplifica isso. Passamos para a função um número (no caso, 12). Como 12 é reconhecido como uma **string** ele não retorna o valor esperado. Operações envolvendo strings serão abordadas em aulas futuras. 

In [1]:
def primeira():
    var1=input('Entre o número desejado: ')
    print(type(var1))
    print(var1*2)

primeira()

Entre o número desejado: 12
<class 'str'>
1212


**Obs.:** Como podemos observar, os valores manipulados em um programa podem ser de diferentes **tipos** ou **classes**. Até agora, vimos que um valor pode ser um número inteiro (classe **int**) ou um texto (classe **str**). A função `type()` nos dá essa informação.

___

Agora, vamos modificar a função `primeira()` para que ela realize a operação desejada envolvendo o número passado no `input()`. Então, precisamos de alguma função que converta uma string (**str**) em um número inteiro (**int**). Para isso, usamos a função de conversão `int()`.

In [2]:
def primeira():
    var1=int(input('Entre o número desejado: '))
    print(type(var1))
    print(var1*2)

In [3]:
primeira()

Entre o número desejado: 12
<class 'int'>
24


In [4]:
type(var1)

NameError: name 'var1' is not defined

Aqui, o erro ocorre porque `var1` só existe dentro da função `primeira()`. Assim que termina a função, ela não existe mais. Por isso, quando executamos `type(var1)` ocorre um erro dizendo que a variável não está definida. 

* **variáveis locais** existem apenas no escopo da função em que foram definidas. 


* **variáveis globais** existem fora da função e podem ser acessadas por quaisquer funções.

## Precedência das Operações

| Nível     	|         Categoria         	| Operadores (em Python)                                           	|
|-----------	|:-------------------------:	|------------------------------------------------------------------	|
| 7 (Alto)  	| exponenciação             	| **                                                               	|
| 6         	| multiplicação (e divisão) 	| *, / (divisão comum), // (divisão inteira), % (resto da divisão) 	|
| 5         	| adição (e subtração)      	| +, -                                                             	|
| 4         	| relacional                	| ==, !=, <=, >= ,<, >                                             	|
| 3         	| lógico                    	| not                                                              	|
| 2         	| lógico                    	| and                                                              	|
| 1 (Baixo) 	| lógico                    	| or                                                               	|

### 1. Operadores aritméticos

Na ausência de parêntesis, as operações são realzadas de acordo com a precedência entre operadores.

In [5]:
4*2+3**3*4

116

O resultado anterior foi obtido pois as operações foram executadas da seguinte forma:

$$4\times2 + (3^3)\times 4 = 8 + (27\times4) = 8 + 108 = 116$$

In [6]:
# Divisão Inteira
10//3

3

In [7]:
# Divisão Comum
10/3

3.3333333333333335

In [8]:
# Resto da Divisão de 10 por 3 (10 = 3*3 + 1)
10%3

1

Algumas funções pertencem a determinadas bibliotecas e por isso, antes de serem utilizadas precisam ser carregadas.

A função que calcula a raiz quadrada de um número é a `sqrt()`, que pertence à biblioteca math (biblioteca que possui funções matemáticas).

Alguns opções de comandos para importar uma biblioteca:

* `import math` - importa todas as funções da biblioteca `math`. Neste caso, para utilizar a função desejada, precisamos, antes do nome dela, dizer de qual pacote ela pertence: `math.sqrt()`.


* `from math import sqrt` - importa apenas a função `sqrt` da biblioteca `math`.


* `from math import *` - importa todas as funções da biblioteca `math`. No entanto, não precisamos dizer de qual pacote a função pertence. 


No decorrer do curso, vamos abordar mais essas diferenças. 

In [9]:
# Importando todos os módulo da biblioteca math
from math import *

In [10]:
round(sqrt(10),2)

3.16

A função `round(numero,n)` recebe um número qualquer e arredonda ele para $n$ casas decimais.

Um valor do tipo **flutuante** é qualquer número decimal, representado em Python pelo tipo de dado **float**.

Também podemos colocar apenas um número dentro da função `round(numero)`, assim ela irá arredondar o número e devolver um inteiro. Com essse mesmo objetivo, a função `int()` também converte um número flutuante em inteiro. 


**Obs.:** A função `round()` tem uma característica especial: se o número flutuante (**float**) tem igual distância entre o inteiro de cima e o inteiro de baixo, ela arredonda para o número par mais próximo. 

In [11]:
round(sqrt(10))

3

In [12]:
type(round(sqrt(10)))

int

In [13]:
int(sqrt(10))

3

In [14]:
# Arredonda para o número par mais próximo
round(2.5)

2

In [15]:
# Arredonda para o número par mais próximo
round(3.5)

4

In [16]:
# Como calcular outras raízes de um número. Se queremos calcular a raiz b de n, fazemos b**(1/n).
10**(1/3)

2.154434690031884

### 2.  Operadores relacionais

Um **operador Relacional** é todo operador que obtém a relação do membro da esquerda com o membro a sua direita, portanto, para utilizar um operador relacional temos que pensar na seguinte estrutura: 
```Python
 <membro à esquerda> OPERADOR <membro à direita>
```

|    Descrição   | Operador (em Python) |
|:--------------:|:--------------------:|
| Igual          |          ==          |
| Diferente      |          !=          |
| Maior          |          >           |
| Maior ou Igual |          >=          |
| Menor          |           <          |
| Menor ou Igual |          <=          |


**OBbs.:** um teste de igualdade é simétrico, mas uma atribuição não é. Por exemplo, se `a == 7` então `7 == a`. Mas, em Python, a expressão `a = 7` é válida e `7 = a` não é.

### 3. Operadores lógicos

Como vimos anteriormente, **expressões aritméticas**, como $4\times2+3^3\times4$, são formadas por valores e **operadores aritméticos**. De maneira semelhante, **expressões lógicas ou booleanas** são formadas por valores e  **operadores lógicos** tais como `not`, `and` (conectivo de conjunção) e `or` (conectivo de disjunção).

| Descrição 	| Operadores (em Python) 	|
|-----------	|------------------------	|
| not       	| negação                	|
| and       	| conjunção              	|
| or        	| disjunção              	|


## Expressões Lógicas
Existem apenas dois **valores booleanos**: verdadeiro (`True`, em Python) ou falso (`False`, em Python). As iniciais maiúsculas são importantes, uma vez que `true` e `false` não são valores booleanos (lembrando que em Python, há diferença entre maísucula e minúscula). Outra observasão é que os valores booleanos não são strings (note que não são escritos entre aspas). 

In [17]:
print(type(True))
print(type('true'))

<class 'bool'>
<class 'str'>


## Tabela Verdade

### not
| p 	| ~p 	|
|:---:	|:----:	|
| V 	| F  	|
| F 	| V  	|



### and, or e xor (ou exclusivo)

| p 	| q 	| p and q 	| p or q 	| p xor q 	|
|:-:	|:-:	|:-------:	|:------:	|:-------:	|
| V 	| V 	|    V    	|    V   	|    F    	|
| V 	| F 	|    F    	|    V   	|    V    	|
| F 	| V 	|    F    	|    V   	|    V    	|
| F 	| F 	|    F    	|    F   	|    F    	|



**Obs.:** O operador **XOR** (também chamado de *ou exclusivo*  ou *disjunção exclusiva*) é uma operação lógica entre dois operandos que resulta em um valor lógico verdadeiro se e somente se o número de operandos com valor verdadeiro for ímpar. 

In [18]:
a=True
b=True

In [19]:
print('not(a): ', not(a))
print('not(b): ', not(b))
print('a and b: ', a and b)
print('a or b: ', a or b)
print('a xor b: ', a ^ b)

not(a):  False
not(b):  False
a and b:  True
a or b:  True
a xor b:  False


### not (p or q)

| p 	| q 	| p or q 	| not (p or q) 	|
|:-:	|:-:	|:------:	|:------------:	|
| V 	| V 	|    V   	|       F      	|
| V 	| F 	|    V   	|       F      	|
| F 	| V 	|    V   	|       F      	|
| F 	| F 	|    F   	|       V      	|

### not (p) and not (q)

| p 	| q 	| not (p) 	| not (q) 	| not(p) and not (q) 	|
|:-:	|:-:	|:-------:	|:-------:	|:------------------:	|
| V 	| V 	|    F    	|    F    	|          F         	|
| V 	| F 	|    F    	|    V    	|          F         	|
| F 	| V 	|    V    	|    F    	|          F         	|
| F 	| F 	|    V    	|    V    	|          V         	|

$$ \frac{C}{5} = \frac{F-32}{9}$$

**EXERCÍCÍO 1:** Vamos fazer uma função que converte a temperatura de de Celsius para Fahrenheit.

Logo, isolando o $F$, temos que: $F = C \times \frac{9}{5} + 32$.

In [20]:
def cf(tempc):
    tempf=9*tempc/5+32
    return tempf

In [21]:
cf(35)

95.0

**EXERCÍCÍO 2:** Agora, vamos fazer uma função que converte a temperatura de Fahrenheit para Celsius.

Logo, isolando o $C$, temos que: $C = 5 \times \frac{F-32}{9}$.

In [22]:
def fc(tempf):
    tempc=5*(tempf-32)/9
    return tempc

In [23]:
fc(95)

35.0

**EXERCÍCÍO 3:** Dada uma temperatura em Fahrenheit, queremos saber como está a temperatura em Celsius.

Aqui, a função `taquente(temp)` recebe uma temperatura em Fahrenheit e utilizamos a função `fc(tempf)`, já criada anteriormente, para converter a temperatura para Celsius. Em seguida, ela realiza um teste condicional por meio do `if...else` para decidir como está a temperatura. 

In [24]:
def taquente(temp):
    tconv=fc(temp)
    if tconv >= 30:
        print('Tá quente demais!')
        return tconv
    else:
        print('Tá na boa!')
        return tconv

In [25]:
taquente(95)

Tá quente demais!


35.0

## Condicionais

**OBS.:** Cuidado com as indentações!

### 1. `if`

Executa o bloco de comandos tabulados (indentados) se a condição é verdadeira.

```Python
if condiçao:
    |
    | bloco de comandos
    |
```

Note que o uso do `else` não é obrigatório.


### 2. `if-else`

Se a condição for verdadeira executa apenas o bloco de comandos 0. Se a condição for falsa executa apenas o bloco de comandos 1.

```Python
if condição:
    |
    | bloco de comandos 0
    |
else:
    |
    | bloco de comandos 1
    |
```

### 3. `if-elif-else`

As condições são testadas em ordem, uma após outra. Apenas o bloco de comandos correspondente a primeira condição que for verdadeira será executada. O bloco de comandos associados ao else serão executados apenas se todas as condições forem falsas. Não é necessário que haja um else ao final.


```Python
if condição_0:
    |
    | bloco de comandos 0
    |
elif condiçao_1:
    |
    | bloco de comandos 1
    |
elif condicão_2:
    |
    | bloco de comandos 2
    |

elif ...:

elif condição_k:
    |
    | bloco de comandos k
    |
else:
    |
    | bloco de comandos k+1
    |
```